### PR TITLE
More data reuse in differentiation

### LINKED ISSUE #144: `_compute_local_gradient` largely duplicates the code of `_apply_stiffness_transpose_operator`
As a result, (versions of) #143 and #142 exist here, too. I think it should be straightforward to make them the same routine with an additional parameter or two.

This is something I should have caught during review of #74.

cc @thomasgibson @lukeolson

### PR DESCRIPTION
**Pull request by inducer**
Still to do:

- [x] Needs https://github.com/inducer/arraycontext/pull/87
- [x] Allow premultipling metric derivatives with jacobian
- [x] Pull xyz_axis into loopy kernel in grad
- [x] Pull xyz_axis into loopy kernel in weak grad
- [x] Unify weak and strong derivatives

cc @kaushikcfd 

Closes #144.

### CODE REVIEW

**Comment by inducer** [comment]
@kaushikcfd This is done and headed for `main` once CI passes. I'll merge it into #161 once it's on main. It may be of interest because it avoids loading the area elements in the differentiation entirely: they get premultiplied into the geometric factors. The same optimization applies for affine, but it likely won't make as much of a difference.

cc @thomasgibson: This cleans up the derivative operators a bit. It avoids the repeated `stack`, unifies back-end implementations of weak and strong, and facilitates premultiplication of the area elements into the inverse metric derivatives.

**Comment by inducer** [comment]
@thomasgibson Please do take a look and leave a review, even if it's post-merge. (It's probably easiest to read this commit-by-commit.)

**Comment by inducer** [comment]
I'll actually leave this off auto-merge for the moment, until someone tells me that the `metric_in_matvec` flag from the last commit isn't insane.

**Comment by inducer** [comment]
Upon sleeping on it, I think it's not too crazy. Still interested in what you think, but I'll likely merge this if it passes.

File: grudge/op.py
Comment (thomasgibson) [inline grudge/op.py:142]: Maybe emphasize here in this function (as a comment or dosctring, nothing too complicated) that this routine is more or less supporting two types of differentiation (weak and strong --- depending on what gets passed in as `get_diff_mat` and `metric_in_matvec`.

**Review by thomasgibson — APPROVED** [review]
I'm overall fine with this. Just a minor comment.

**Comment by thomasgibson** [comment]
> Upon sleeping on it, I think it's not too crazy. Still interested in what you think, but I'll likely merge this if it passes.

I don't think it's too crazy at all. The only difference is a scaling factor in the geometric terms and the elemental matrix. It's the same type of local mat vec otherwise, so I'm totally cool with this.

**Comment by inducer** [comment]
Hi all, I totally fumbled the social aspect of this. My attempts to figure out whether I should just merge this or wait on a review may have ended up creating a sense of urgency that just didn't exist. I apologize if any of you felt pressured in any way, that was not the intention. Sorry!

File: grudge/op.py
Comment (inducer) [reply grudge/op.py]: Agreed, see 3987a0a.

**Comment by inducer** [comment]
Thanks for taking a look @thomasgibson!

**Pull request merged**
